src/Core/Domain/Order/ValueObject/OrderDetailRefund.php

```diff
@@ -0,0 +1,143 @@
+<?php
+/**
+ * 2007-2020 PrestaShop SA and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2020 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Order\ValueObject;
+
+use PrestaShop\PrestaShop\Core\Domain\Order\Exception\EmptyRefundAmountException;
+use PrestaShop\PrestaShop\Core\Domain\Order\Exception\EmptyRefundQuantityException;
+use PrestaShop\PrestaShop\Core\Domain\Order\Exception\OrderException;
+
+/**
+ * Class ProductRefund
+ */
+class OrderDetailRefund
+{
+    /**
+     * @var int
+     */
+    private $orderDetailId;
+
+    /**
+     * @var int
+     */
+    private $productQuantity;
+
+    /**
+     * @var float|null
+     */
+    private $refundAmount;
+
+    /**
+     * @param int $orderDetailId
+     * @param int $productQuantity
+     * @param float $refundAmount
+     *
+     * @return self
+     *
+     * @throws EmptyRefundAmountException
+     * @throws OrderException
+     */
+    public static function createPartialRefund(int $orderDetailId, int $productQuantity, float $refundAmount): self
+    {
+        if (0 >= $refundAmount) {
+            throw new EmptyRefundAmountException();
+        }
+
+        return new self($orderDetailId, $productQuantity, $refundAmount);
```

Comment:
Is this method really needed? Why not using the constructor, and add the check above in a validator method?

Reply:
It's not mandatory, but the idea is to close the constructor access, and use static functions which clearly define the context because this whole feature is a bit difficult to understand

A partial refund MUST have a defined amount, just like a standard refund has NO defined amount (it is the product price by design), so I wanted the available code to translate these ideas

This could be done through different ValueObject classes as well (`OrderDetailPartialRefund extends OrderDetailStandardRefund` or vice versa), but it seemed overkill